components/steps/nz-step.component.ts

```diff
@@ -1,52 +1,33 @@
-import {
-  Component,
-  ElementRef,
-  Input,
-  TemplateRef,
-  ViewChild
-} from '@angular/core';
+import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, OnChanges, SimpleChanges, TemplateRef, ViewChild } from '@angular/core';
 
-import { NzUpdateHostClassService } from '../core/services/update-host-class.service';
+import { classMapToString } from '../core/style/map';
 import { NgClassType } from '../core/types/ng-class';
 
 @Component({
+  changeDetection    : ChangeDetectionStrategy.OnPush,
   selector           : 'nz-step',
-  providers          : [ NzUpdateHostClassService ],
   preserveWhitespaces: false,
-  templateUrl        : './nz-step.component.html'
+  templateUrl        : './nz-step.component.html',
+  host               : {
+    '[class]': 'classString'
+  }
 })
-export class NzStepComponent {
-  private _status = 'wait';
-  private _currentIndex = 0;
-  private _description: string | TemplateRef<void>;
-  private _icon: NgClassType | TemplateRef<void>;
-  private _title: string | TemplateRef<void>;
-  private el: HTMLElement = this.elementRef.nativeElement;
-  oldAPIIcon = true; // Make the user defined icon compatible to old API. Should be removed in 2.0.
-  isCustomStatus = false;
-  isDescriptionString = true;
-  isTitleString = true;
-  isIconString = true;
-  last = false;
-  showProcessDot = false;
-  direction = 'horizontal';
-  outStatus = 'process';
-  index = 0;
+export class NzStepComponent implements OnChanges {
   @ViewChild('processDotTemplate') processDotTemplate: TemplateRef<void>;
-  customProcessTemplate: TemplateRef<{ $implicit: TemplateRef<void>, status: string, index: number }>;
 
   @Input()
+  get nzTitle(): string | TemplateRef<void> { return this._title; }
   set nzTitle(value: string | TemplateRef<void>) {
     this.isTitleString = !(value instanceof TemplateRef);
     this._title = value;
   }
-
-  get nzTitle(): string | TemplateRef<void> {
-    return this._title;
-  }
+  isTitleString = true;
+  private _title: string | TemplateRef<void>;
 
   @Input()
+  get nzIcon(): NgClassType | TemplateRef<void> { return this._icon; }
```

Comment:
all setter should be after getter
https://angular.io/guide/component-interaction#intercept-input-property-changes-with-a-setter